### ISSUE TITLE
Add "build script" option

### ISSUE THREAD

**Issue by andrerom** [composer/satis#153]
*tl;dr; Like PRE_ARCHIVE_CMD.  However after download, before archive generation, and in a way that Satis can reuse it.*

## Feature
As a Satis User I need to be able to execute build logic before archive is made, this involves for instance:
- generate custom file md5 checksum for files in the archive (to provide GUI info on changed files)
- change copyright/license files in case of distribution under several license
- Anything else people tend to do in build scripts on CI systems

This seems to be needed at archive stage to make sure satis/composer reliably generates packages with same checksum each time, as opposed to new checksum on each run in case you do these changes by rewriting git tags locally before executing satis.

## Possible Approach(es)
Currently investigating how to execute events like post-install-cmd before package is archived by composer. (Unfortunately) It seems that all this logic is executed in ArchiveManager->archive(), so options seems to be:
- Add parameter to inject optional $eventName + eventDispatcher in ctor, and call before $usableArchiver->archive()
- refactor ArchiveManager->archive() into two functions, one for download and one for archive to be able to use these from satis to perform the event dispatching entirely in satis
- other approach

@Seldaek @naderman Any suggestions/pov on how to accomplish this in a way that can be accepted as PR(s)?

**Comment by vdespa** [comment]
@andrerom I am having a similar problem and I am curious about your conclusions.

Is this proposal abandoned? 

Did you find any better solutions that worked for you? Thanks.

**Comment by andrerom** [comment]
Currently stuck using a fork to have this, you can find it here:
- https://github.com/ezsystems/satis/commits/ez
- This relies on a couple of changes to composer: https://github.com/andrerom/composer/commits/ez

Besides the change to composer to compress archives *(which I'm surprised not more people ask for, maybe I overlook something)*, the build step stuff is something we intend to try to rather do before invoking satis, some low level features of git does allow you to exchange commits without changing the commit sha1, meaning satis will be able to build several variants with same sha1.

**Comment by vdespa** [comment]
@andrerom Thanks for your reply.

We decided to go with another solution. We use the CI tool (Jenkins in our case) to do the extra build stuff that we need and generate an artefact.

We link the artefact in the composer.json

    "dist": {
        "url": "http://ci-server:8080/job/test/450/artifact/dummy/**/*zip*/archive.zip",
        "type": "zip"
    }

This way it is possible to request via composer just the source file or the dist. 

The only back-draw is that you need to change / commit the composer.json every-time you do a build.

**Comment by alcohol** [comment]
> compress archives 

You want to compress something that is already compressed?


Anyway, not sure what the state of this issue is. Not really clearly to me what is asked for, nor if this still applies to the current codebase. Closing for now. Feel free to request a re-open if you have some reason that merits it.
